library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,568 @@
+      res["cleanup"] = cleanup_method if respond_to?(:cleanup)
+      res["widget"] = widget_type
+
+      res
+    end
+
+    # gets if widget is open for modification
+    def enabled?
+      Yast::UI.QueryWidget(Id(widget_id), :Enabled)
+    end
+
+    # Opens widget for modification
+    def enable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, true)
+    end
+
+    # Closes widget for modification
+    def disable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, false)
+    end
+
+  protected
+
+    # helper to check if event is invoked by this widget
+    def my_event?(event)
+      widget_id == event["ID"]
+    end
+
+    # shortcut from Yast namespace to avoid including whole namespace
+    # TODO: kill converts in CWM module, to avoid this workaround for funrefs
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    # @note all methods here use wrappers to modify required parameters as CWM
+    # have not so nice callbacks API
+    def init_method
+      fun_ref(method(:init_wrapper), "void (string)")
+    end
+
+    def init_wrapper(_widget)
+      init
+    end
+
+    def handle_method
+      fun_ref(method(:handle_wrapper), "symbol (string, map)")
+    end
+
+    # allows both variant of handle. with event map and without.
+    # with map it make sense when handle_all_events is true or in custom widgets
+    # with multiple elements, that generate events, otherwise map is not needed
+    def handle_wrapper(_widget, event)
+      m = method(:handle)
+      if m.arity == 0
+        m.call
+      else
+        m.call(event)
+      end
+    end
+
+    def store_method
+      fun_ref(method(:store_wrapper), "void (string, map)")
+    end
+
+    def store_wrapper(_widget, _event)
+      store
+    end
+
+    def cleanup_method
+      fun_ref(method(:cleanup_wrapper), "void (string)")
+    end
+
+    def cleanup_wrapper(_widget)
+      cleanup
+    end
+
+    def validate_method
+      fun_ref(method(:validate_wrapper), "boolean (string, map)")
+    end
+
+    def validate_wrapper(_widget, _event)
+      validate
+    end
+  end
+
+  # Represents custom widget, that have its UI content defined in method content.
+  # Useful mainly when specialized widget including more subwidget should be
+  # reusable at more places.
+  #
+  # @example custom widget child
+  #   class MyWidget < CWM::CustomWidget
+  #     def initialize
+  #       self.widget_id = "my_widget"
+  #     end
+  #
+  #     def contents
+  #       HBox(
+  #         PushButton(Id(:reset), _("Reset")),
+  #         PushButton(Id(:undo), _("Undo"))
+  #       )
+  #     end
+  #
+  #     def handle(widget, event)
+  #       case event["ID"]
+  #       when :reset then ...
+  #       when :undo then ...
+  #       else ...
+  #       end
+  #     end
+  #   end
+  class CustomWidget < AbstractWidget
+    self.widget_type = :custom
+    # custom witget without contents do not make sense
```

Comment:
But we need to document `contents`. Just declaring it abstract is not enough.
>\# @return [Term] ...

Is the term allowed to contain AbstractWidgets, or just plain UI terms?